### ISSUE TITLE
Introduce dynamic cluster management in Ambry

### ISSUE THREAD

**Issue by pnarayanan** [linkedin/ambry#524]
Cluster management is static in Ambry today. Nodes read in the hardware and partition layouts from files. Any layout change requires an update to the files, deployment and restarts of the nodes. This is not a scalable approach and is an impediment to more advanced features that involve dynamic layout changes, such as automatic handling of disk and node failures.

This is a high-level issue to track the investigation, design and implementation of a dynamic, coordinated cluster management option in Ambry.

**Comment by pnarayanan** [comment]
The investigation and recommendation to use [Apache Helix](https://github.com/apache/helix) for cluster management is [here](https://docs.google.com/document/d/1H5jZwIHBf-92xjGAY2q2J6acSH9cA30rdCiCWQTopsQ).

**Comment by pnarayanan** [comment]
The design for dynamic cluster management using Helix is [here](https://docs.google.com/document/d/1gMweKKzpgcGciXzhNpjI3gf9973QhkZoYA6YkUhaFvU).

**Comment by siepkes** [comment]
If I understand it correctly this issue about replacing the static configuration files for Ambry with something more dynamic. I've read the recommendation document which speaks of solutions like Zookeeper and Helix on one end and incorporating the RAFT algorithm on the other end. I was curious as to why the "middle ground" is not discussed in the document; Off loading it to distributed solutions like Hashicorp's `Consul` and CoreOs's `etcd`  (which use RAFT)?

Zalando for example uses this construct in [Patroni](https://github.com/zalando/patroni) (distributed PostgreSQL). They also put a small abstraction (not more then an basic interface) in front of the implementation which allows for a broad support of implementations (Consul, etcd, zookeeper, etc.). The reason why this might be interesting is because a lot of sites usually have some sort of distributed  solution like Zookeeper, Consul, etc. already deployed.

**Comment by pnarayanan** [comment]
@siepkes Thank you for looking into the docs and for your comments. We looked into Zookeeper and Helix as far as the coordination-as-a-service solutions are concerned, primarily because those are the systems that were and are readily available and supported for use within our sites (at LinkedIn). We looked at RAFT to see if we can avoid any dependencies altogether. I am not sure if the solutions you described are really a "middle ground" however, but I have not looked into those in detail.

As far as the [implementation](https://github.com/linkedin/ambry/pull/569) within Ambry is concerned, we have the abstractions in place so that agents of any coordination service - which follows or can be made to follow the model of cluster constituents registering themselves to announce that they are alive, and cluster viewers subscribing for change notifications - can be plugged in (with a little effort).
